admiral/envs/components/actor.py

```diff
@@ -0,0 +1,248 @@
+
+    position (ContinuousPositionState):
+        The position state handler. Needed to modify agent positions.
+    
+    speed_angle (SpeedAngleState):
+        The speed and angle state handler.
+
+    agents (dict):
+        The dictionary of agents.
+    """
+    def __init__(self, position=None, speed_angle=None, agents=None, **kwargs):
+        self.position = position
+        self.speed_angle = speed_angle
+        self.agents = agents
+
+        from gym.spaces import Box
+        for agent in agents.values():
+            if isinstance(agent, SpeedAngleAgent):
+                agent.action_space['accelerate'] = Box(-agent.max_acceleration, agent.max_acceleration, (1,))
+                agent.action_space['banking_angle_change'] = Box(-agent.max_banking_angle_change, agent.max_banking_angle_change, (1,))
+    
+    def move(self, agent, acceleration, angle, **kwargs):
+        self.speed_angle.modify_speed(agent, acceleration)
+        self.speed_angle.modify_banking_angle(agent, angle)
+        
+        x_position = agent.speed*np.cos(np.deg2rad(agent.ground_angle))
+        y_position = agent.speed*np.sin(np.deg2rad(agent.ground_angle))
+        
+        position_before = agent.position
+        self.position.modify_position(agent, np.array([x_position, y_position]))
+        return agent.position - position_before
+
+
+
+# -------------------------------- #
+# --- Resources and Harvesting --- #
+# -------------------------------- #
+
+class GridResourcesActor:
+    """
+    Provides the necessary action space for agents who can harvest resources and
+    processes the harvesting action.
+
+    resources (ResourceState):
+        The resource state handler.
+
+    agents (dict):
+        The dictionary of agents.
+    """
+    def __init__(self, resources=None, agents=None, **kwargs):
+        self.resources = resources
+        self.agents = agents
+
+        from gym.spaces import Box
+        for agent in agents.values():
+            if isinstance(agent, HarvestingAgent):
+                agent.action_space['harvest'] = Box(0, agent.max_harvest, (1,), np.float)
+
+    def process_harvest(self, agent, amount, **kwargs):
+        """
+        Harvest some amount of resources at the agent's position.
+
+        agent (HarvestingAgent):
+            The agent who has chosen to harvest the resource.
+
+        amount (float):
+            The amount of resource the agent wants to harvest.
+        
+        return (float):
+            Return the amount of resources that was actually harvested. This can
+            be less than the desired amount if the cell does not have enough resources.
+        """
+        if isinstance(agent, HarvestingAgent) and isinstance(agent, PositionAgent):
+            location = tuple(agent.position)
+            resource_before = self.resources.resources[location]
+            self.resources.modify_resources(location, -amount)
+            return resource_before - self.resources.resources[location]
```

Comment:
To be consistent with movement, we should return the state after - the state before.

Reply:
No, because the resource is being depleted, so I want to have a positive amount that represents how much was harvested.